llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -1886,6 +1886,119 @@ LogicalResult PackOp::generateScalarImplementation(OpBuilder &builder,
+  bindDims(ctx, dim0, dim1);
+  auto subMap = AffineMap::get(2, 0, {dim0 - dim1});
+  auto sub = [&](OpFoldResult v1, OpFoldResult v2) {
+    return builder.createOrFold<AffineApplyOp>(
+        loc, subMap,
+        ValueRange{getValueOrCreateConstantIndexOp(builder, loc, v1),
+                   getValueOrCreateConstantIndexOp(builder, loc, v2)});
+  };
+
+  int64_t inputRank = getInputRank();
+  DenseMap<int64_t, OpFoldResult> dimAndTileMapping = getDimAndTileMapping();
+  SmallVector<OpFoldResult> sourceIndices, sourceSizes;
+  for (auto dim : llvm::seq<int64_t>(0, inputRank)) {
+    if (dimAndTileMapping.count(dim)) {
+      AffineExpr i, tile;
+      bindDims(ctx, i);
+      bindSymbols(ctx, tile);
+      OpFoldResult sourceIndex = makeComposedFoldedAffineApply(
+          builder, loc, i * tile,
+          ArrayRef<OpFoldResult>{offsets[dim], dimAndTileMapping[dim]});
+      sourceIndices.push_back(sourceIndex);
+
+      OpFoldResult sourceSize = makeComposedFoldedAffineApply(
+          builder, loc, i * tile,
+          ArrayRef<OpFoldResult>{sizes[dim], dimAndTileMapping[dim]});
+
+      sourceSizes.push_back(sourceSize);
+    } else {
+      sourceIndices.push_back(offsets[dim]);
+      sourceSizes.push_back(sizes[dim]);
+    }
+    OpFoldResult dimSize = getDim(builder, loc, getInput(), dim);
+    sourceSizes.back() =
+        min(sourceSizes.back(), sub(dimSize, sourceIndices.back()));
+  }
+
+  auto oneAttr = builder.getI64IntegerAttr(1);
+  SmallVector<OpFoldResult> strides(inputRank, oneAttr);
+
+  SmallVector<Value> tiledOperands;
+  tiledOperands.push_back(
+      getSlice(builder, loc, getInput(), sourceIndices, sourceSizes, strides));
+
+  SmallVector<OpFoldResult> outputOffsets, outputSizes;
+  if (failed(getResultTilePosition(builder, 0, offsets, sizes, outputOffsets,
+                                   outputSizes))) {
+    return {};
+  }
+  strides.append(inputRank, oneAttr);
+  tiledOperands.push_back(
+      getSlice(builder, loc, getOutput(), outputOffsets, outputSizes, strides));
+
+  for (auto tile : getInnerTiles()) tiledOperands.push_back(tile);
+  if (auto val = getPaddingValue()) tiledOperands.push_back(val);
+
+  // There are exactly one input and one output, the output is the second operand.
+  SmallVector<Type, 4> tiledResultTypes;
+  if (hasTensorSemantics()) {
+    tiledResultTypes.push_back(tiledOperands[1].getType());
+  }
+
+  Operation *tiledPackOp =
+      cast<LinalgExtOp>(getOperation())
+          .clone(builder, loc, tiledResultTypes, tiledOperands);
+
+  return {tiledPackOp};
+}
+
+LogicalResult PackOp::getResultTilePosition(
+    OpBuilder &builder, unsigned resultNumber, ArrayRef<OpFoldResult> offsets,
+    ArrayRef<OpFoldResult> sizes, SmallVector<OpFoldResult> &resultOffsets,
+    SmallVector<OpFoldResult> &resultSizes) {
+  auto zeroAttr = builder.getI64IntegerAttr(0);
+  resultOffsets.assign(offsets.begin(), offsets.end());
+  resultOffsets.append(getInputRank(), zeroAttr);
```

Comment:
I think this should be `getOutputRank() - getInputRank()` (instead of just `getInputRank()`)